### DISCUSSION TITLE
Derived / Computed query selector

### DISCUSSION THREAD

**Discussion by talberkoMelio** [TanStack/query#3961] [Ideas]
Hi
I would like to suggest a new feature - derived (computed) query selector.
The selector will be able to "observe" one or more queries, and export a derived value.
Is is something thats in your roadmap? I would like to contribute - please let me know how to get started if it's relevant.

Thanks

**Comment by TClark1011** [comment]
How would this be different from the `select` option you can use in `useQuery`?

> select: (data: TData) => unknown
> Optional
> This option can be used to transform or select a part of the data returned by the query function. It affects the returned data value, but does not affect what gets stored in the query cache.

**Reply by talberkoMelio** [reply]
@TClark1011 ot has 2 main differences:
1. Select is for the query it was added to, not for multiple queries (for example in order for data aggregation).
2. The select will be triggered only if the query is mounted. (opposing to QueryObserver).

So I want to combine these 2.

**Reply by TClark1011** [reply]
I understand your first point, but not the second, what do you mean by the select only running when the query is mounted?

**Reply by talberkoMelio** [reply]
Let's assume I have a query I use in some route by using useQuery. This useQuery will have a selector that returns the data. 
Mounting the route will trigger data fetching.
Before fetching is done, I will move to another route (which will unmount the useQuery hook from above).
The data will be settled in the query and be cached, but selector will not run. Not sure if this is a bug or in purpose.

**Reply by TkDodo** [reply]
> The data will be settled in the query and be cached, but selector will not run. Not sure if this is a bug or in purpose.

That is on purpose and not sure why it is a problem. The data that is cached is the one that is returned from the `queryFn`. Then, everyone who wants to read the data can optionally transform it with `select` and thus create a more fine-grained subscription.

So if the second route wants to read the data, it can also do this with `useQuery` and potentially call `select` on it. Maybe you can elaborate more on you example because I don't understand the use-case yet.

**Reply by talberkoMelio** [reply]
@TkDodo the issue here, is that I would expect the selector to trigger when data is settled, and not only if the useQuery is currently rendered.
The current behavior - I render a useQuery with a selector (that inside I log "hello"). Then before data is settled (the api call is sent, but not yet done), I unmount the useQuery. No "hello" is printed, but data is settled.
Now when I use the query again, data is not transformed.

**Reply by talberkoMelio** [reply]
@TkDodo I will explain my intent, and my suggested solution:
I have 3 different queries. These queries fetch lists data, and when all are settled (or changed), I want to group them all.
The current flow I implemented, is having 3 queries, each query has queryObserver subscribed to is, and when data is settled, I set the data in recoil.
In recoil, I use recoil selector getter as my computed/derived data for grouped items.

Instead of doing this, I would like to suggest adding the option for derived/computed/getter in reactQuery.

I think this can be a great addition to reactQuery.

WDYT?

**Reply by TkDodo** [reply]
> Now when I use the query again, data is not transformed.

If you use the query again, use it with the same selector. Or, don't use a selector at all, but transform in the queryFn. The differences are outlined here: https://tkdodo.eu/blog/react-query-data-transformations

> The current flow I implemented, is having 3 queries, each query has queryObserver subscribed to is, and when data is settled, I set the data in recoil.

there are already multiple ways to do this:
- you can do one query that makes 3 api calls and then transform in the queryFn. This is good if you don't need the queries separately.
- you can `useQueries` and then just transform the resulting data array once you have all 3 values. This is good if you need the single query hooks separately as well.

**Reply by talberkoMelio** [reply]
@TkDodo I need all 3 separately, and also need to hold these 3 as one grouped object.
I see I can't use the `useQueries` and pass a `UseQueryResult`, or just a key. So do I need to pass the same object as I pass to the single `useQuery` (key and queryFn)?

**Reply by talberkoMelio** [reply]
@TkDodo BTW, I was trying to use onSettled function, so when data is settled, I want to store it in the recoil state.
I noticed that when the data is settled, but the query is not mounted, the onSettled is not triggered - also not in the next time it is mounted.
(I added log to the onSettled. I see it prints only when data is settled while the query is mounted, otherwise, data is settled in the query, but the function is not triggered). 
Is it a bug?

**Reply by TkDodo** [reply]
> So do I need to pass the same object as I pass to the single useQuery (key and queryFn)?

yes.

> I noticed that when the data is settled, but the query is not mounted, the onSettled is not triggered

callbacks are on a per-observer level. if the observer (your component) unmounts, there is no callback to call anymore. If you mount the component twice, both `onSettled` callbacks will be called.

It's generally not a good idea to do state syncing from the callbacks.

**Reply by talberkoMelio** [reply]
@TkDodo I think using `useQueries` will not solve the problem, as its code will run each time I will render it.
For example I will use useQueries to orchestrate all 3 indivitual queries into one, and return a grouped object. The grouped object is not cached similar to the useQuery itself.
I would like to have a way to access a "computed" data, which depends on multiple queries, and is cached similar to the useQuery hooks. Any way to acheive it using the current api of react-query?

**Reply by TkDodo** [reply]
why wouldn't useQueries solve it?

```
const [result1, result2] = useQueries(query1, query2)

const computedResult = React.useMemo(() => {
  if (result1.data && result2.data) return computeResult(result1.data, result2.data)
  return null
}, [result1.data, result2.data])
```

downside is that a component that uses this might render a bit more often, because results will come in asynchronously and there is no way to bail out of rendering at that point. But this is likely not a problem, and if it is, pass `computedResult` to a memoized component and do the rest there, so the parts that render often are not expensive.

---

Other than that, you can always build a single query that does multiple requests with `Promise.allSettled`, and then use `select` to transform the data.

But no, anything else for deriving data from multiple queries is not supported.

**Reply by non25** [reply]
It's kinda sad to do this in react land. :slightly_smiling_face: 
Each component that uses such derivation will make another clone, unless you push it to the context above, which de-opt it even further by triggering re-renders.

Currently going through discussions on the repo. :grin: 
Posted some relevant thoughts [here](https://github.com/TanStack/query/discussions/2178#discussioncomment-6432574)

**Comment by talberkoMelio** [comment]
Any update from moderators?
